ios/ConvosNSE/NotificationService.swift

```diff
@@ -0,0 +1,283 @@
+      } else {
+           os_log("Missing required fields (possibly ethAddress?) or incorrect type in userInfo/body dictionary.", log: logger, type: .error)
+      }
+
+      contentHandler?(request.content); return
+    }
+
+    os_log("Processing message - Topic: %{public}@, EthAddress: %{public}@", log: logger, type: .info, topic, ethAddress) // This will log the lowercased one now
+
+    // --- 2. Retrieve Configuration & Keys ---
+    guard let xmtpEnvString = getInfoPlistValue(key: infoPlistXmtpEnvKey) else {
+      os_log("Failed to read XMTP environment from Info.plist", log: logger, type: .error)
+      contentHandler?(request.content); return
+    }
+    let xmtpEnv = getXmtpEnvironmentFromString(envString: xmtpEnvString)
+
+    guard let messageBytes = Data(base64Encoded: Data(encryptedMessage.utf8)) else {
+      os_log("Failed to decode base64 encryptedMessage payload", log: logger, type: .error)
+      contentHandler?(request.content); return
+    }
+
+    // --- 3. Build XMTP Client ---
+    // Use shared App Group directory for the database - using the dynamic group ID
+    guard let groupDir = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: currentAppGroupId)?.path else {
+      os_log("Failed to get App Group container URL for group %{public}@", log: logger, type: .error, currentAppGroupId)
+      contentHandler?(request.content); return
+    }
+
+    let options = ClientOptions(
+      api: .init(env: xmtpEnv, isSecure: true), // Assuming isSecure=true for dev/prod
+      dbEncryptionKey: dbEncryptionKeyData, // Use the key data retrieved earlier
+      dbDirectory: groupDir
+    )
+
+    let identity = PublicIdentity(kind: .ethereum, identifier: ethAddress) // Use address determined earlier
+
+    do {
+      // Note: Building the client might be resource-intensive for an NSE. Monitor performance.
+      let client = try await Client.build(publicIdentity: identity, options: options)
+      os_log("XMTP client built successfully for group %{public}@", log: logger, type: .info, currentAppGroupId) // Changed to .info
+
+
+      // --- 4. Decrypt Message ---
+      os_log("Attempting to find conversation by topic: %{public}@", log: logger, type: .info, topic) // Changed to .info
+
+      guard let conversation = try? await client.conversations.findConversationByTopic(topic: topic) else {
+        os_log("Conversation not found for topic: %{public}@", log: logger, type: .default, topic)
+        currentBestAttempt.title = "New Message"
+        currentBestAttempt.body = "[Encrypted]"
+        contentHandler?(currentBestAttempt); return
+      }
+      os_log("Conversation found. Syncing...", log: logger, type: .info) // Changed to .info
+      try? await conversation.sync()
+
+      os_log("Processing message bytes...", log: logger, type: .info) // Changed to .info
+      guard let decodedMessage = try? await conversation.processMessage(messageBytes: messageBytes) else {
```

Comment:
You might be able to get away with doing this optimistically. Try and process the message first, then if it fails sync and fetch the message by ID. 

Most of the time processMessage will work without needing the sync. Only will fail if you are an epoch behind.